### PR TITLE
[RUNTIME] Add TypedPackedFunc

### PR DESCRIPTION
**Pull request by tqchen**
TypedPackedFunc enables compile-time type checking. TypedPackedFunc works with the runtime system:
- It can be passed as an argument of PackedFunc,
- It can be assigned to TVMRetValue.
- It can be directly converted to a type-erased PackedFunc.

This allows us to write more typed code in c++ while taking benefit of PackedFunc that can be used interchangeably with frontend code.

## Example
```c++
void TestTypedPackedFunc() {
  using Int1Func = TypedPackedFunc<int(int)>;
  using Int2Func = TypedPackedFunc<int(int, int)>;
  using BindFunc = TypedPackedFunc<Int1Func(Int2Func, int value)>;
  BindFunc ftyped;
  ftyped = [](Int2Func f1, int value) -> Int1Func {
    auto binded = [f1, value](int x) {
      return f1(value, x);
    };
    Int1Func x(binded);
    return x;
  };
  auto add = [](int x, int y) { return x + y; };
  CHECK_EQ(ftyped(Int2Func(add), 1)(2), 3);
  PackedFunc f = ftyped(Int2Func(add), 1);
  CHECK_EQ(f(3).operator int(), 4);
  // call the type erased version.
  Int1Func f1 = ftyped.packed()(Int2Func(add), 1);
  CHECK_EQ(f1(3), 4);
}
```

### CODE REVIEW

**Comment by tqchen** [comment]
@jroesch @yzhliu  please review

File: include/tvm/runtime/packed_func.h [hunk 1]
```diff
@@ -118,6 +118,163 @@ class PackedFunc {
...
+class TypedPackedFunc;
+
+/*!
+ * \anchor TypedPackedFuncAnchor
+ * \brief A PackedFunc wrapper to provide typed function signature.
+ * It is backed by a PackedFunc internally.
+ *
+ * TypedPackedFunc enables compile time type checking.
+ * TypedPackedFunc works with the runtime system:
+ * - It can be passed as an argument of PackedFunc,
```
Comment (jroesch) [inline include/tvm/runtime/packed_func.h]: trailing comma should be a period?

File: include/tvm/runtime/packed_func.h [hunk 2]
```diff
@@ -118,6 +118,163 @@ class PackedFunc {
...
+ * - It can be assigned to TVMRetValue.
+ * - It can be directly converted to a type-erased PackedFunc.
+ *
+ * Please use it when possible in the c++ code to enable more
```
Comment (jroesch) [inline include/tvm/runtime/packed_func.h]: Maybe something like "Developers should prefer TypedPackedFunc over PackedFunc in C++ code as it enables compile time checking."

**Comment by yzhliu** [comment]
Are we going to use it in frontend language bindings?

**Comment by tqchen** [comment]
@yzhliu This is mainly for c++ backend code, in the frontend binding, what user see is mainly PackedFunc. This being said, we can also try to support type signature for strongly typed language like java, if we have ways to dynamically build up the function

**Comment by tqchen** [comment]
To follow up on what this might be helpful, currently, nnvm uses std::function for most of the operator attributes, which makes it hard to interchange between frontend(python) and backend. We want to upgrade the operator attribute system so that next version we can directly use PackedFunc, but we still want to have the std::function signature so that the c++ side of pass gives you strong type and compiler will hit an error when you mistakenly call it with wrong signature

**Comment by yzhliu** [comment]
got it. thanks for explaining.

**Pull request merged**